src/main/java/org/launchcode/water_garden_tour/models/garden/GardenData.java

```diff
@@ -0,0 +1,102 @@
+
+    public static List<Garden> gardenSearchFeaturesFiltered(String searchValueRaw, Integer[] featureIds, Iterable<Feature> allFeatures, List<Garden> allGardens) {
+        String searchValue = searchValueRaw.toLowerCase();
+        List<Garden> results = new ArrayList<>();
+        Feature curFeature;
+
+        for (Garden garden : allGardens) {
+            boolean hasAllFeatures = true;
+
+            for (Integer featureId : featureIds) {
+                curFeature = getFeatureFromId(featureId, allFeatures);
+
+                if (garden.getFeatures().contains(curFeature)) {
+                } else {
+                    hasAllFeatures = false;
+                }
+            }
+            if (hasAllFeatures) {
+                results.add(garden);
+            }
+        }
+
+        if (searchValue.equals("all") || searchValue.equals("")){
+
+        } else {
+            List<Garden> iteratingGardenList = new ArrayList<>();
+            iteratingGardenList.addAll(results);
+
+            for (Garden garden : iteratingGardenList) {
+                if (!isTextInGardenFields(garden, searchValue)) {
+                    results.remove(garden);
+                }
+            }
+
+        }
+        return results;
+    }
+
+    public static List<Garden> gardenSearchAllFeatures(String searchValueRaw, Iterable<Feature> allFeatures, List<Garden> allGardens) {
+        String searchValue = searchValueRaw.toLowerCase();
+        List<Garden> results = new ArrayList<>();
+
+        if (searchValue.equals("all") || searchValue.equals("")){
+            results = allGardens;
+        } else {
+
+        }
+            for (Garden garden : allGardens) {
+
+                for(Feature feature : garden.getFeatures()) {
+
+                    if (feature.getName().toLowerCase().contains(searchValue)) {
+
+                        if (!results.contains(garden)) {
+
+                            results.add(garden);
+                        }
+                    }
+                }
+
+                if (isTextInGardenFields(garden, searchValue)) {
+                    if (!results.contains(garden)) {
+                        results.add(garden);
+                    }
+                }
+        }
+
+        return results;
+    }
+
+    public static Feature getFeatureFromId(Integer featureId, Iterable<Feature> allFeatures) {
+
+        for (Feature feature : allFeatures) {
+            if (featureId.equals(feature.getId())) {
+                return feature;
+            }
+        }
+        return null;
+    }
+
+    public static boolean isTextInGardenFields(Garden curGarden, String searchValue) {
+
+            if (curGarden.getName().toLowerCase().contains(searchValue) ||
+                    curGarden.getDescription().toLowerCase().contains(searchValue) ||
+                    curGarden.getAddress().toLowerCase().contains(searchValue) ||
+                    curGarden.getOwner().getName().toLowerCase().contains(searchValue)) {
```

Comment:
because this call is doing a lot I might create a method in the Garden class to do this and then call that here. Just an idea you don't have to

Reply:
I was using this chunk of code a few times in my search methods above so I decided to make it it's own method. I put it here in GardenData because it gets used here but could put it in Garden if that is more appropriate.

Reply:
nope here is fine just wanted to know more about it good work!